x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/HashAggregationOperator.java

```diff
@@ -192,15 +208,29 @@ public void finish() {
         try {
             selected = blockHash.nonEmpty();
             Block[] keys = blockHash.getKeys();
-            int[] aggBlockCounts = aggregators.stream().mapToInt(GroupingAggregator::evaluateBlockCount).toArray();
-            blocks = new Block[keys.length + Arrays.stream(aggBlockCounts).sum()];
-            System.arraycopy(keys, 0, blocks, 0, keys.length);
-            int offset = keys.length;
-            for (int i = 0; i < aggregators.size(); i++) {
-                var aggregator = aggregators.get(i);
-                aggregator.evaluate(blocks, offset, selected, driverContext);
-                offset += aggBlockCounts[i];
+
+            int blockCount = 0;
+            for (int g = 0; g < groups.size(); g++) {
+                blockCount += groups.get(g).finishBlockCount();
+            }
+            int[] aggBlockCounts = new int[aggregators.size()];
+            for (int a = 0; a < aggregators.size(); a++) {
+                aggBlockCounts[a] = aggregators.get(a).evaluateBlockCount();
+                blockCount += aggBlockCounts[a];
```

Comment:
I found it a lot easier to read if I encoded the `resultOffset`s into the GroupKeys. It'd be even easier to read if the offsets were encoded into the aggregators too. Or if we returns `Block[]`.